File: runners/go/llog.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"fmt"

	"gopkg.in/yaml.v3"
)

// LLog is the parsed logical op-log (the YAML authoring profile of
// spec/l-log.schema.json). Only the subset the Phase 0-2 corpus needs is
// modeled; unknown fields are ignored rather than rejected so the model can
// grow without breaking existing fixtures.
type LLog struct {
	Header  Header  `yaml:"header"`
	Entries []Entry `yaml:"entries"`
}

// Header is the determinism contract + initial table definition.
type Header struct {
	ID            string            `yaml:"id"`
	FormatVersion int               `yaml:"format-version"`
	SpecAnchor    string            `yaml:"spec-anchor"`
	TableUUID     string            `yaml:"table-uuid"`
	Schema        LSchema           `yaml:"schema"`
	PartitionSpec *LPartitionSpec   `yaml:"partition-spec"`
	Properties    map[string]string `yaml:"properties"`
}

// LSchema is the op-log schema block: an ordered list of fields.
type LSchema struct {
	Fields []LField `yaml:"fields"`
}

// LField is one schema field. Type is a raw yaml.Node because an iceberg type
// is either a primitive string (e.g. "long", "decimal(9,2)") or a nested type
// object (struct/list/map); schema.go resolves it.
type LField struct {
	ID             int         `yaml:"id"`
	Name           string      `yaml:"name"`
	Type           yaml.Node   `yaml:"type"`
	Required       bool        `yaml:"required"`
	Doc            string      `yaml:"doc"`
	InitialDefault *TypedValue `yaml:"initial-default"`
	WriteDefault   *TypedValue `yaml:"write-default"`
}

// LPartitionSpec is the header partition-spec block.
type LPartitionSpec struct {
	SpecID int               `yaml:"spec-id"`
	Fields []LPartitionField `yaml:"fields"`
}

// LPartitionField is one partition field with a transform (e.g. "bucket[4]").
type LPartitionField struct {
	SourceID  int    `yaml:"source-id"`
	FieldID   int    `yaml:"field-id"`
	Name      string `yaml:"name"`
	Transform string `yaml:"transform"`
}

// Entry is one op-log entry. Op discriminates; the remaining fields are the
// union of every op's payload (only the fields relevant to Op are populated).
type Entry struct {
	Op string `yaml:"op"`

	// append / overwrite
	Rows []Row `yaml:"rows"`

	// delete / overwrite
	Predicate      *Predicate `yaml:"predicate"`
	Strategy       string     `yaml:"strategy"`
	Kind           string     `yaml:"kind"`
	EqualityFields []int      `yaml:"equality-fields"`

	// evolve-schema
	Changes []yaml.Node `yaml:"changes"`

	// evolve-spec
	Spec *LPartitionSpec `yaml:"spec"`

	// observe
	At   yaml.Node `yaml:"at"`
	Bind string    `yaml:"bind"`

	// authored-golden / oracle payloads are for the orchestrator; the runner
	// parses but does not act on them.
	Expect    yaml.Node `yaml:"expect"`
	Invariant yaml.Node `yaml:"invariant"`
}

// Row is one logical row: a required synthetic __rowkey plus field values
// keyed by column name, each value a TypedValue carrying its physical type.
type Row struct {
	RowKey string
	Values map[string]*TypedValue
}

// UnmarshalYAML decodes a row mapping, peeling off __rowkey and decoding every
// other entry as a TypedValue (which understands the !int/!long/... tags).
func (r *Row) UnmarshalYAML(node *yaml.Node) error {
	if node.Kind != yaml.MappingNode {
		return fmt.Errorf("row must be a mapping, got kind %d", node.Kind)
	}
	r.Values = map[string]*TypedValue{}
	for i := 0; i+1 < len(node.Content); i += 2 {
		key := node.Content[i].Value
		val := node.Content[i+1]
		if key == "__rowkey" {
			r.RowKey = val.Value
			continue
		}
		tv := &TypedValue{}
		if err := tv.fromNode(val); err != nil {
			return fmt.Errorf("row field %q: %w", key, err)
		}
		r.Values[key] = tv
	}
	if r.RowKey == "" {
		return fmt.Errorf("row missing __rowkey")
	}
	return nil
}

// Predicate mirrors the l-log predicate grammar reduced to what the corpus
// uses. Only the eq form (term + typed value) is needed for Phase 2.
type Predicate struct {
	Type   string        `yaml:"type"`
	Term   string        `yaml:"term"`
	Value  *TypedValue   `yaml:"value"`
	Values []*TypedValue `yaml:"values"`
	Args   []*Predicate  `yaml:"args"`
	Arg    *Predicate    `yaml:"arg"`
}

File: runners/rust/src/llog.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! Parsed logical op-log (the YAML authoring profile of spec/l-log.schema.json).
//!
//! The outer structure is decoded with serde derive; per-value physical types
//! (the `!int` / `!long` / `!struct` YAML tags) are preserved as raw
//! `serde_yaml::Value` and interpreted by [`crate::values`], since serde's
//! derive cannot dispatch on YAML tags directly.

use serde::Deserialize;
use serde_yaml::Value as YamlValue;

/// The whole op-log: determinism/DDL header + ordered DML entries.
#[derive(Debug, Deserialize)]
pub struct LLog {
    pub header: Header,
    pub entries: Vec<Entry>,
}

/// Determinism contract + initial table definition. Unknown fields are ignored
/// so the model can grow without breaking existing fixtures. `partition_spec`
/// is parsed but not yet applied (partition-spec support is Phase 4 work).
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct Header {
    #[serde(default)]
    pub id: String,
    #[serde(rename = "format-version", default)]
    pub format_version: i32,
    /// "synthesized" (default) builds the table from schema+entries; "artifact"
    /// loads a checked-in table read-only (read conformance).
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub artifact: Option<LArtifact>,
    /// Required for source=synthesized; absent for read fixtures.
    #[serde(default)]
    pub schema: Option<LSchema>,
    #[serde(rename = "partition-spec", default)]
    pub partition_spec: Option<YamlValue>,
    #[serde(default)]
    pub properties: std::collections::HashMap<String, String>,
}

/// Locates a checked-in table for read (source: artifact) mode.
#[derive(Debug, Deserialize)]
pub struct LArtifact {
    /// Fixture-relative dir holding bytes/ + bytes/ROOT.
    pub path: String,
}

/// The op-log schema block: an ordered list of fields.
#[derive(Debug, Deserialize)]
pub struct LSchema {
    pub fields: Vec<LField>,
}

/// One schema field. `field_type` is a raw YAML value because an iceberg type
/// is either a primitive string ("long", "decimal(9,2)") or a nested
/// struct/list/map object; [`crate::schema`] resolves it.
#[derive(Debug, Deserialize)]
pub struct LField {
    pub id: i32,
    pub name: String,
    #[serde(rename = "type")]
    pub field_type: YamlValue,
    #[serde(default)]
    pub required: bool,
    #[serde(default)]
    pub doc: Option<String>,
    #[serde(rename = "initial-default", default)]
    pub initial_default: Option<YamlValue>,
    #[serde(rename = "write-default", default)]
    pub write_default: Option<YamlValue>,
}

/// One op-log entry. `op` discriminates; the remaining fields are the union of
/// every op's payload (only those relevant to `op` are populated).
///
/// Some fields are parsed (so serde accepts the fixtures) but not yet read by
/// the emit-only runner: `predicate`/`strategy`/`changes` are consumed once
/// delete/evolve leave declared-gap status; `expect`/`invariant` are always
/// orchestrator-only.
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct Entry {
    pub op: String,

    // append: rows is a sequence of row mappings (kept raw for tagged values).
    #[serde(default)]
    pub rows: Option<YamlValue>,

    // delete
    #[serde(default)]
    pub predicate: Option<YamlValue>,
    #[serde(default)]
    pub strategy: Option<String>,
    #[serde(default)]
    pub kind: Option<String>,

    // evolve-schema
    #[serde(default)]
    pub changes: Option<YamlValue>,

    // observe: 'latest', a bound name, or a commit ordinal.
    #[serde(default)]
    pub at: Option<YamlValue>,
    #[serde(default)]
    pub bind: Option<String>,

    // authored-golden / oracle payloads are for the orchestrator; parsed but
    // not acted on by the emit-only runner.
    #[serde(default)]
    pub expect: Option<YamlValue>,
    #[serde(default)]
    pub invariant: Option<YamlValue>,
}
